In [1]:
import pandas as pd

In [2]:
train_df= pd.read_csv('../input/arkon-data-bikes-train/train_set.csv')
test_df= pd.read_csv('../input/arkon-data-bikes-test/test_set.csv')

/opt/conda/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3552: DtypeWarning: Columns (8) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


# 2.- Data Preparation

In [3]:
train_df.dtypes

trip_id                  int64
duration                 int64
start_time              object
end_time                object
start_lat              float64
start_lon              float64
end_lat                float64
end_lon                float64
bike_id                 object
plan_duration          float64
trip_route_category     object
passholder_type         object
start_station            int64
end_station              int64
dtype: object

In [4]:
train_df.dtypes

trip_id                  int64
duration                 int64
start_time              object
end_time                object
start_lat              float64
start_lon              float64
end_lat                float64
end_lon                float64
bike_id                 object
plan_duration          float64
trip_route_category     object
passholder_type         object
start_station            int64
end_station              int64
dtype: object

In [5]:
train_df.isnull().sum()

trip_id                    0
duration                   0
start_time                 0
end_time                   0
start_lat               5563
start_lon               5563
end_lat                18574
end_lon                18574
bike_id                    0
plan_duration            208
trip_route_category        0
passholder_type         2576
start_station              0
end_station                0
dtype: int64

In [6]:
test_df.isnull().sum()

trip_id                    0
duration                   0
start_time                 0
end_time                   0
start_lat               4622
start_lon               4622
end_lat                14891
end_lon                14891
bike_id                    0
trip_route_category        0
start_station              0
end_station                0
dtype: int64

In [7]:
print(len(train_df.columns))
print(len(test_df.columns))

14
12


Como es posible observar, en ambos conjuntos de datos (train_set y test_set) existen valores nulos en diferentes columnas. Para este problema existen dos opciones de solución: eliminar los ejemplos con datos faltantes o "rellenarlos" (***Data Imputation*** ) con nuevos valores.

Eliminar los datos se presenta como la opción más rápida y sencilla. Sin embargo, esto implicaría la perdida de ejemplos que probablemente afectarían el rendimiento de cualquier modelo.

Por lo anterior se decidió utilizar ***Data Imputation*** a partir del cálculo de la media aritmética con el objetivo de evitar el problema descrito previamente y simplificar el proceso de aplicación de esta técnica.

Sin embargo, es necesario eliminar 2576 ejemplos del conjunto de entrenamiento debido a que no esta presnete el valor de 'passholder_type'.

In [8]:
# Review columns
NoMatch_columns = []
for column in train_df.columns:
    for index, column2 in enumerate(test_df.columns):
        if column == column2:
            break
        elif index + 1 == len(test_df.columns):
            NoMatch_columns.append(column)
NoMatch_columns

['plan_duration', 'passholder_type']

In [9]:
train_df.isnull().sum()

trip_id                    0
duration                   0
start_time                 0
end_time                   0
start_lat               5563
start_lon               5563
end_lat                18574
end_lon                18574
bike_id                    0
plan_duration            208
trip_route_category        0
passholder_type         2576
start_station              0
end_station                0
dtype: int64

### passholder_type

In [10]:
len(train_df)

700000

In [11]:
train_df = train_df.dropna(subset = ['passholder_type'])

In [12]:
len(train_df)

697424

### plan_duration

In [13]:
train_df.drop('plan_duration', inplace = True, axis = 1)

### start_time and end_time

In [14]:
columns = ['start_time', 'end_time']
for column in columns:
    train_df[column] = pd.to_datetime(train_df[column])
    test_df[column] = pd.to_datetime(test_df[column])

In [15]:
train_df['start_time'].head(6)

0   2018-08-07 11:20:00
1   2017-09-17 17:51:00
2   2019-04-22 09:22:00
3   2019-09-22 11:27:00
4   2020-01-31 17:11:00
5   2017-12-16 15:18:00
Name: start_time, dtype: datetime64[ns]

In [16]:
test_df['start_time'].head(6)

0   2017-01-01 00:24:00
1   2017-01-01 00:28:00
2   2017-01-01 00:39:00
3   2017-01-01 00:43:00
4   2017-01-01 00:56:00
5   2017-01-01 01:54:00
Name: start_time, dtype: datetime64[ns]

### trip_id

In [17]:
train_df.drop('trip_id', inplace = True, axis = 1)
test_df.drop('trip_id', inplace = True, axis = 1)

### start_lat, start_lon, end_lat, end_lon

In [18]:
dict_temp = {'start_lat':0, 'start_lon':0, 'end_lat':0, 'end_lon':0}

for column in dict_temp.keys():
    dict_temp[column] = train_df[column].mean()
    train_df[column] = train_df[column].fillna(dict_temp[column])
    test_df[column] = test_df[column].fillna(dict_temp[column])

### bike_id

### trip_route_category

### passholder_type

In [19]:
train_df.isnull().sum()

duration               0
start_time             0
end_time               0
start_lat              0
start_lon              0
end_lat                0
end_lon                0
bike_id                0
trip_route_category    0
passholder_type        0
start_station          0
end_station            0
dtype: int64

In [20]:
train_df.isnull().sum()

duration               0
start_time             0
end_time               0
start_lat              0
start_lon              0
end_lat                0
end_lon                0
bike_id                0
trip_route_category    0
passholder_type        0
start_station          0
end_station            0
dtype: int64

In [21]:
print(len(train_df.columns))
print(len(test_df.columns))

12
11


In [22]:
train_df.head()

duration          start_time            end_time  start_lat   start_lon  \
0        35 2018-08-07 11:20:00 2018-08-07 11:55:00  33.748920 -118.275192   
1        32 2017-09-17 17:51:00 2017-09-17 18:23:00  34.035679 -118.270813   
2         6 2019-04-22 09:22:00 2019-04-22 09:28:00  34.046070 -118.233093   
3       138 2019-09-22 11:27:00 2019-09-22 13:45:00  34.062580 -118.290092   
4        14 2020-01-31 17:11:00 2020-01-31 17:25:00  34.026291 -118.277687   

     end_lat     end_lon bike_id trip_route_category passholder_type  \
0  33.748920 -118.275192   06530          Round Trip         Walk-up   
1  34.047749 -118.243172    6683             One Way         Walk-up   
2  34.047749 -118.243172   06710             One Way    Monthly Pass   
3  34.059689 -118.294662   17068             One Way    One Day Pass   
4  34.021660 -118.278687   18841             One Way    Monthly Pass   

   start_station  end_station  
0           4127         4127  
1           3057         3062  
2           3022         3062  
3           4304         4311  
4           4266         4443

In [23]:
test_df.head()

duration          start_time            end_time  start_lat   start_lon  \
0        12 2017-01-01 00:24:00 2017-01-01 00:36:00  34.058319 -118.246094   
1        17 2017-01-01 00:28:00 2017-01-01 00:45:00  34.049980 -118.247162   
2        20 2017-01-01 00:39:00 2017-01-01 00:59:00  34.063389 -118.236160   
3        12 2017-01-01 00:43:00 2017-01-01 00:55:00  34.048851 -118.246422   
4        48 2017-01-01 00:56:00 2017-01-01 01:44:00  34.049198 -118.252831   

     end_lat     end_lon bike_id trip_route_category  start_station  \
0  34.058319 -118.246094    6351          Round Trip           3028   
1  34.043732 -118.260139    5836             One Way           3027   
2  34.044159 -118.251579    6529             One Way           3066   
3  34.050140 -118.233238    6029             One Way           3029   
4  34.049198 -118.252831    6680          Round Trip           3063   

   end_station  
0         3028  
1         3018  
2         3055  
3         3079  
4         3063

Falta TRANFORMAR VARIABLES CATEGORICAS

## Save Files

In [24]:
train_df.to_csv('train_set_final.csv', index=False)
test_df.to_csv('test_set_final.csv', index=False)